### PR TITLE
Add min_p sampling support

### PR DESCRIPTION
**Pull request by DwieDave**
This PR adds support for min_p sampling, which helps control token sampling diversity 
by filtering out tokens with probabilities below min_p multiplied with the probability of the most likely token. 
This feature is useful for improving text coherence while maintaining some randomness, 
particularly in long-context chatting scenarios with smaller models.

Related issue: #4

Key changes:
1. API integration:
- Add `min_p` parameter to completions API endpoint (defaults to 0.0)
- Create dummy `".*"` regex grammar when `min_p > 0.0` to ensure logit processing
- Pass `min_p` through `AsyncExecutor::add_request()`
- Calculate `ln(min_p)` in `PendingSeq::step()` for active requests

2. Core sampling implementation:
- Add `ln_min_p` field to `TlcLogitsEntry` struct (initialized to `-FLT_MAX`)
- Extend CUDA kernel to filter tokens below `ln_min_p` threshold using `logit_adjusted` comparison
- Adjusted host memory allocation to accommodate `ln_min_p` values

Please let me know if there are any changes necessary to approve this PR.

### CODE REVIEW

File: llgtrt/src/routes/completions.rs [hunk 1]
```diff
@@ -282,7 +282,16 @@ fn llg_grammar(params: &CommonCreateParams) -> Result<Option<TopLevelGrammar>> {
             log::debug!("using Lark grammar");
             lark_to_llguidance(lark_grammar)?
         }
-        _ => return Ok(None),
+        _ => {
+            if params.min_p > 0.0 {
+                // Returning a Dummy-grammar to enforce logit processing when min_p is set
+                let grm = TopLevelGrammar::from_regex(llguidance::api::RegexNode::Regex(
+                    ".*".to_string(),
```
Comment (mmoskal) [inline llgtrt/src/routes/completions.rs]: You should at least use `r"(\n|.)*"` which will allow newlines. However, it will still not allow special tokens. I need to think about the most efficient way to allow them all.

File: trtllm-c/mask_logits.cu [hunk 2]
```diff
@@ -135,6 +135,10 @@ __global__ void mask_logits_kernel(T** logit_ptrs, int64_t* mask_offsets, size_t
             else
             {
                 logit_adjusted = (logit - s_max_val_allowed) * beta;
+                if ((float) logit_adjusted < ln_min_p[batch_idx])
+                {
+                    logit_adjusted = -flt_max;
```
Comment (mmoskal) [inline trtllm-c/mask_logits.cu:140]: What happens when there isn't any logit larger than the limit?

File: llgtrt/src/routes/openai.rs [hunk 3]
```diff
@@ -1,7 +1,7 @@
 use super::api_ext::LlgLogLevel;
 use llguidance::api::TopLevelGrammar;
 use serde::{Deserialize, Deserializer, Serialize};
-use std::collections::HashMap;
+use std::{collections::HashMap, f32};
```
Comment (mmoskal) [inline llgtrt/src/routes/openai.rs]: no need to import f32

File: llgtrt/src/routes/openai.rs [hunk 4]
```diff
@@ -165,6 +165,9 @@ pub struct CommonCreateParams {
     /// tokens comprising the top 10% probability mass are considered.
     #[serde(default = "default_top_p")]
     pub top_p: f32,
+    /// Filters out tokens with probability less than min_p multiplied by the probability of the most likely token
+    #[serde(default = "default_min_p")]
```
Comment (mmoskal) [inline llgtrt/src/routes/openai.rs]:
```suggestion
    #[serde(default)]
```

Default default :) is `0.0`

File: trtllm-c/logits.cpp [hunk 5]
```diff
@@ -201,8 +207,8 @@ static void logitsPostProcessorFn(std::vector<tle::IdType> const& reqIds, std::v
     if (dp > 0)
     {
         cudaMemcpyAsync(cudaMasksData, masksData, max_offset + mask_stride, cudaMemcpyHostToDevice, stream);
-        mask_logits_ext(cudaLogitPtrs, cudaMasksOffsets, mask_fractions_offset, temperatures_offset, dp, nVocab,
-            mask_stride / 4, tp, stream);
+        mask_logits_ext(cudaLogitPtrs, cudaMasksOffsets, mask_fractions_offset, temperatures_offset, ln_min_p_offset,
```
Comment (mmoskal) [inline trtllm-c/logits.cpp:210]: we should probably pass an array of structs with mask offset, temp, and min_p - it's getting too many for comfort ; but you can keep it as is for now

**Review by mmoskal — COMMENTED** [review]
Overall looks great, thank you very much!

Would be good to move min_p into params struct to avoid it being passed separately.

File: trtllm-c/mask_logits.cu (hunk 2 above)
Comment (mmoskal) [reply trtllm-c/mask_logits.cu:140]: sorry, I didn't read, ignore

**Comment by DwieDave** [comment]
@mmoskal Thank you for the review! 
I just pushed a commit that addresses your comments. 
Maybe I will check out the refactor to the array of structs at a later point in time.

**Comment by mmoskal** [comment]
Thanks, I take a final look later today!

**Comment by mmoskal** [comment]
Thank you!
